File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/control/path/PathPoint.java
package org.firstinspires.ftc.teamcode.control.path;

import android.annotation.SuppressLint;

import org.firstinspires.ftc.teamcode.control.system.Functions;
import org.firstinspires.ftc.teamcode.util.Point;

public class PathPoint extends Point {
    public double followDistance;
    public String signature;
    public Functions.Function func;

    public PathPoint(String signature, double x, double y, double followDistance) {
        this(signature, x, y, followDistance, null);
    }
    public PathPoint(String signature, double x, double y, double followDistance, Functions.Function func) {
        super(x, y);
        this.signature = signature;
        this.followDistance = followDistance;
        this.func = func;
    }

    public PathPoint(PathPoint b) {
        this(b.signature, b.x, b.y, b.followDistance, b.func);
    }

    @SuppressLint("DefaultLocale")
    @Override
    public String toString() {
        return String.format("%s, %.1f, %.1f, %.1f", signature, x, y, followDistance);
    }

    public boolean equals(PathPoint b) {
        return x == b.x && y == b.y;
    }
}//sdd

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/control/path/PathPoint.kt
package org.firstinspires.ftc.teamcode.control.path

import org.firstinspires.ftc.teamcode.util.Point
import org.firstinspires.ftc.teamcode.util.Pose

// id love for this to be a dataclass but yeah sucks to suck
open class PathPoint(
    val signature: String = "",
    val x: Double = 0.0,
    val y: Double = 0.0,
    val followDistance: Double = 0.0,
    val func: Functions.Function? = null
) {
    val p get() = Point(x, y)

    open val copy: PathPoint get() = PathPoint(signature, x, y, followDistance, func)
    fun distance(p2: PathPoint) = p.distance(p2.p)
    fun distance(p2: Pose) = p.distance(p2.p)

    override fun toString(): String {
        return String.format(
            "%s, %.1f, %.1f, %.1f, %s",
            signature,
            x,
            y,
            followDistance,
            func
        )
    }
}
